Source/Core/AudioCommon/CubebUtils.cpp

```diff
@@ -62,6 +66,11 @@ std::shared_ptr<cubeb> CubebUtils::GetContext()
     ERROR_LOG(AUDIO, "Error setting cubeb log callback");
   }
 
+#ifdef _WIN32
+  // Cubeb requires us to do it ourselves
+  CoInitializeEx(nullptr, COINIT_MULTITHREADED);
```

Comment:
Note that this is a leak, of sorts (not sure what happens if you keep calling it without uninitializing, might just increment a counter).

It can also fail (probably won't, but still good practice to check the return value).

Microsoft documentation claims `CoInitializeEx must be called at least once, and is usually called only once, for each thread that uses the COM library`, which is *not* satisfied by just calling it in `GetContext()`. The returned context can and will be used across multiple threads. *However,* in practice this isn't an issue, as once the multithreaded apartment is initialized, all threads without COM initialized will act as though they in the multithreaded apartment. More information in kinetiknz/cubeb#534

Reply:
Good find, I even went on to check every line of code in Cubeb to make sure it didn't have memory leaks and forgot to check on my own code. This would be a leak when the game/emulation thread is closed but Dolphin is not.

CoInitializeEx can be called as many times as you want on the same thread, calls after the first one have no effect, but it will need to be uninitialised yeah (if any of the calls succeeded).

I'm wondering if we need to call CoInitialize ourselves on the audio thread as well, given that cubeb_wasapi.cpp never calls it anywhere. Though it probably wouldn't work without it so I guess it's fine as it is.

Reply:
> Good find, I even went on to check every line of code in Cubeb to make sure it didn't have memory leaks and forgot to check on my own code. This would be a leak when the game/emulation thread is closed but Dolphin is not.
>
> CoInitializeEx can be called as many times as you want on the same thread, calls after the first one have no effect, but it will need to be uninitialised yeah (if any of the calls succeeded).

Right, it's not a leak per-se, it's just you won't be uninitializing it so I wanted to call it out. Probably worth putting comments in explaining what's happening.

This can affect multiple threads: whichever threads are creating the CubebStream (system thread? UI thread?), as well as whichever thread is creating the EXI_DeviceMic (system thread?). I remember seeing that it's very easy to trigger this. Again, not really a problem, but since it's counter to COM best practice I wanted to point it out. It would be needlessly convoluted to set it up correctly in Dolphin, its really cubeb's responsibility (I have an old cubeb PR open for this).

> I'm wondering if we need to call CoInitialize ourselves on the audio thread as well, given that cubeb_wasapi.cpp never calls it anywhere. Though it probably wouldn't work without it so I guess it's fine as it is.

This goes back to the default-multithreaded-apartment-once-the-multithreaded-apartment-is-created implementation: if any thread, anywhere, has called CoInitializeEx, then you can use COM from anywhere. Not documented, but works in practice.

Reply:
Should be fixed, please take a look

Reply:
I need to make sure only one thread can create and access the cubeb backend. Regarding the GC controller microphone, it should be fine. I actually forgot earlier but every succeeding call to CoInitializeEx need to have a CoUninitialize yeah, it probably works with a counter internally.

Reply:
Oh nope, that won't work. Since cubeb doesn't keep a reference to COM, uninitializing after get a cubeb context will break.

Like I said, there's not really a great way to handle this in Dolphin. I'd just call CoInitializeEx every time you need a cubeb context, punt on uninitializing, and put in a comment explaining the issue.

Reply:
The CoUninitialize happens on the destruction of the cubeb context, which happens after the stream is closed, when the audio backend shuts down, so it's perfectly fine.

Reply:
You're right, my mistake. I didn't see the if-clause around the second CoUninitializeEx.

What happens if a cubeb context is created on one thread, and destroyed on another? Is COM okay with that?

Reply:
No it wouldn't be ok, but it doesn't happen, the audio backend is created and destroyed from the same thread.
What is definitely called from multiple threads is SoundStream::SetRunning(bool running), whether Cubeb or WASAPI (I know because I had troubles with the wasapi CoInit there).
Also, not sure about the microphone code threading, I will ask around since I can't test it.

Reply:
When I tested this last time, it does get called from multiple threads, and causes segfaults.

Reply:
https://github.com/dolphin-emu/dolphin/blob/92d24aec2e45fc65767444b220108595f8e74b94/Source/Core/InputCommon/ControllerInterface/Win32/Win32.cpp#L49 this line here is breaking WASAPI CoInit calls because some are COINIT_APARTMENTTHREADED and some are COINIT_MULTITHREADED. Are you fine with me changing that one to COINIT_MULTITHREADED?

Reply:
No actually, that doesn't seem the problem. I'm getting RPC_E_CHANGED_MODE but I always use COINIT_MULTITHREADED. Investigating

Reply:
It seems like QT or some other library is calling CoInitializeEx with COINIT_APARTMENTTHREADED in the main thread, then, my WASAPI calls use COINIT_MULTITHREADED so i get the RPC_E_CHANGED_MODE. I'm trying to understand which one we actually want to use then.

Reply:
The microphone access seems to be safe and only done from one thread (I haven't tested at all but it seems that way)

Reply:
Microphone access from multiple threads can be reproduced by:
1. starting the game with a microphone configured in Slot A or B
2. unconfiguring the microphone while the game is running

Reply:
True. I didn't know it was that easy to test. Not a problem, I will initialize and unitialize COM every time a cubeb function is called from a dolphin thread.

Reply:
The fact that you can receive RPC_E_CHANGED_MODE is irrelevant as we can just keep going, unfortunately, there isn't much we can do, it's likely an external library that sets it up as COINIT_APARTMENTTHREADED .
Before this conversion, cubeb itself used to be COINIT_MULTITHREADED, and we likely want to keep it that way as all the code from our side thread safe.

Reply:
> Not a problem, I will initialize and unitialize COM every time a cubeb function is called from a dolphin thread.
 
This might work sometimes, but isn't how COM is supposed to be use. Any objects living in the apartment will be destroyed when the apartment is uninitialized. If you happen to get one of those, it will segfault the next time it's used.

Reply:
If that's so, it would have done the same before this Cubeb update. The only change is that CoInit calls are moved before calling Cubeb instead than being inside of it.

Reply:
But given that an apartment is already created even before lunching a game (when querying stuff for WASAPI in the UI and calling CoInit, I get RPC_E_CHANGED_MODE), then it's likely that whoever called the first CoInit, won't call CoUnit until the very end of the main thread. We can't guarantee that, but we can't do anything else, it's not in Dolphin or any externals (unless there are Windows calls that wrap CoInit inside of them).

Reply:
And I doubt the apartment would be uninitialised until the very last CoUnit, so we have control over that, given that we have our main cubeb context CoInit ongoing until it is destroyed when the audio backend is shut down or the microphone removed.

Reply:
@ligfx Do you know if accessing a COM object created in a STA thread from an MTA thread is fine? That's when you get the RPC_E_CHANGED_MODE.
Now, cubeb in Dolphin currently has this "problem" already so it's nothing new, again, all that changed is the place where CoInitializeEx is called, but the old cubeb ignores the RPC_E_CHANGED_MODE error, and just keeps going.
I'm still new to this stuff, all I can see is that to access a COM object created in a STA thread from another STA thread, you need marshalling, but I can't find if it's fine to access it from an MTA thread, and whether you need to marshal it yourself, or if it does it for you internally (I don't know what it means exactly). All I know is that all of this is thread safe and there are no concurrent accesses.
Also, note that the STA thread won't be destroyed nor call CoUninit (mirroring the first CoInit) until the very end of the life of the application, so that is not a problem.

Reply:
I believe all WASAPI COM objects live in the multithreaded apartment regardless of where they are created, so you can access them from STA or MTA threads fine (they already assume multithreaded access, unlike objects in the STA).

Reply:
In WASAPI CoCreateInstance() of the IMMDeviceEnumerator is created on a MTA thread and accessed from both MTA and STA threads, but the IAudioClient can be created from an MTA and STA thread (main or emulation threads), and that is also a COM object.
In cubeb, it's very similar, the same can happen.

Reply:
> I believe all WASAPI COM objects live in the multithreaded apartment regardless of where they are created, so you can access them from STA or MTA threads fine (they already assume multithreaded access, unlike objects in the STA).

By the way, not at all, the current implement of wasapi calls CoInitialize, not CoInitializeEx, which implies they are in an STA. But I have completely re-written the WASAPI stream, as it had loads of crashes and non thread safe accesses.

Reply:
I don't think you understood what I said. Let's step back a bit.

In COM, [objects live in certain apartments](https://docs.microsoft.com/en-us/previous-versions/ms809971(v=msdn.10)?redirectedfrom=MSDN), depending on their declared threading model. Threading models are "Apartment" aka STA-only, "Free" aka free-threaded aka MTA-only, or "Both", aka whichever apartment it was created from.

When I say that WASAPI COM objects live in the MTA, I mean that I think they have the "Free" threading model (you can check in the Registry).

So what happens when you create an MTA-only/free-threaded object from an STA? COM will create the MTA if it doesn't exist, and put the object there.

The point being, it doesn't matter which apartment you create or access free-threaded objects from, as long as COM has been initialized.

Reply:
All right. No I had not understood that threading model concept at all, I had misinterpreted, I need to go back and read more. But, other than that, it seems like this PR is good to go then? (I still need to commit a CoInitialize inside CubebStream::SetRunning())

Reply:
What happens if you start a game with the microphone enabled, then disable it while the game is running? My worry is calling CoUninitialize on the wrong thread would break things.

Reply:
Check again, it should be fine now. Again, remember that these problems are currently present already.

Reply:
Yes there is that problem. void CEXIMic::StreamInit() and void CEXIMic::StreamTerminate() can be called by different threads for the same object. I'm not sure how to fix that (though it worked fine for this whole time)

Reply:
A solution could be to wrap all the cubeb calls, like cubeb_init, cubeb_stream_init and cubeb_stream_start around CoInitializeEx and CoUninitialize, that way, no CoInitialize will ever be pending between threads. Will the COM objects be destroyed though?

Reply:
> Will the COM objects be destroyed though?

They might be. Note this is the reason why cubeb decided to move away from just wrapping calls, and towards requiring callers to handle COM: kinetiknz/cubeb#416

Reply:
yes i had read that

Reply:
all right then, what about this:
call the first CoInitialize by creating a thread_local ScopedThreadAccess (see it in CubebUtils.h), then, wrap every call to any cubeb method into another CoInitialize and CoUninitialize. When the thread is destroyed, the thread_local variable will be destroyed and call the final CoUninitialize.

Reply:
thread_local could work.

Not sure if you'd still need to wrap calls: COM seems to treat threads that haven't initialized COM as being in the MTA, if the MTA has already been initialized. Something to test.

Reply:
Microsoft says: every thread that accesses COM objects needs to have called CoInit. If you don't call it, it's treated as implicit MTA: https://devblogs.microsoft.com/oldnewthing/20130419-00/?p=4613

Reply:
Ah, yep, there it is. So as long as at least one thread exists with an MTA, which is probably a reasonable assumption (e.g. emu thread starts before anything else, creates audio and EXI devices), you don't have to wrap calls.

Reply:
Check again, I've done what I mentioned

Reply:
Why avoid relying on the implicit MTA? It makes the code much cleaner, and doesn't expose the need to use COM to users of CubebUtils.

I'd suggest something like this:

```diff
diff --git a/Source/Core/AudioCommon/CubebUtils.cpp b/Source/Core/AudioCommon/CubebUtils.cpp
index b4af87115b..7e500ceb1b 100644
--- a/Source/Core/AudioCommon/CubebUtils.cpp
+++ b/Source/Core/AudioCommon/CubebUtils.cpp
@@ -14,6 +14,10 @@
 
 #include <cubeb/cubeb.h>
 
+#ifdef _WIN32
+#include <Objbase.h>
+#endif
+
 static ptrdiff_t s_path_cutoff_point = 0;
 
 static void LogCallback(const char* format, ...)
@@ -42,8 +46,47 @@ static void DestroyContext(cubeb* ctx)
   }
 }
 
+#ifdef _WIN32
+// cubeb's WASAPI requires the user to manage a valid COM context on all threads
+// accessing cubeb. There are two parts to this: 1) every thread that calls GetContext()
+// will have a COM context created if it doesn't exist 2) while a multi-threaded
+// COM context exists, any thread that doesn't initialize COM will automatically
+// be in the implicit MTA COM context.
+
+class auto_com {
+public:
+  auto_com() = default;
+  ~auto_com() {
+    if (m_should_uninitialize) {
+      CoUninitialize();
+    }
+  }
+  bool initialize() {
+    if (!m_initialized) {
+      HRESULT result = CoInitializeEx(nullptr, COINIT_MULTITHREADED);
+      m_initialized = SUCCEEDED(result) || result == RPC_E_CHANGED_MODE;
+      m_should_uninitialize = SUCCEEDED(result);
+    }
+    return m_initialized;
+  }
+private:
+  bool m_initialized = false;
+  bool m_should_uninitialize = false;
+};
+
+thread_local auto_com s_com_context;
+
+#endif // _WIN32
+
 std::shared_ptr<cubeb> CubebUtils::GetContext()
 {
+  #ifdef _WIN32
+    if (!s_com_context.initialize()) {
+       ERROR_LOG(AUDIO, "Error initializing COM for cubeb");
+       return nullptr;
+    }
+  #endif
+
   static std::weak_ptr<cubeb> weak;
 
   std::shared_ptr<cubeb> shared = weak.lock();
```

Reply:
Did you manage to build PR this on Windows BTW?

Reply:
I think this the above code can work. GetContext() was called from the main thread, it would return RPC_E_CHANGED_MODE, meaning that it won't create any MTA. If there is no MTA, there is also no implicit MTA, right?

Reply:
I don't have a Windows computer at the moment so I can't build.

Here's the logic:
1) Emulation starts, emuthread is created, which starts audio (calls GetContext) and initializes EXI devices (might call GetContext), so initializes an MTA context and ensures the MTA exists.
2) If a microphone is hotplugged, the UI thread will call GetContext and get RPC_E_CHANGED_MODE. No problem, the MTA already exists, and it's fine to create WASAPI objects from the STA due to their threading model. 
3) If a microphone is hotunplugged, the UI thread will call cubeb functions directly. No problem, the MTA exists.
4) Any other thread that does stuff with cubeb: the emuthread should be running, which means the MTA exists.

Reply:
Yes, I kind of understand that it would work, but this makes too many assumptions (some of which I don't have the knowledge to judge).
-How do you know the threading model of all WASAPI COM objects?
-cubeb calls CoInit with MTA on a lot of its functions just for satefy, and it calls it on every new thread, never relying on implicit MTA. Implicit MTA can be used yes, but if the thread that created the MTA closes it when another acesss is ongoing, it's not good. This doesn't happen now, but why risk it to avoid a few CoInit calls?
-What if some other code someone else implements ends up calling CoInit with STA on every relevant thread before we get the chance of opening an MTA? This code makes the assumption that won't happen.

Reply:
> How do you know the threading model of all WASAPI COM objects? Plus, we don't know which objects cubeb uses.

You can look it up in the registry. It's sufficient just to know that Firefox runs cubeb in MTA threads, though.

> Implicit MTA can be used yes, but if the thread that created the MTA closes it when another access is ongoing, it's not good. This doesn't happen now, but why risk it to avoid a few CoInit calls?

Because the scoped CoInit calls don't actually solve anything. In your scenario:
1. At least one other thread has a reference to the MTA
2. We call CoInitializeEx before using cubeb
3. All other threads drop reference to the MTA
4. We call CoUninitialize after using cubeb
5. *Now, the MTA is gone. All COM objects created in the MTA might be invalid and cause crashes.*

> What if some other code someone else implements ends up calling CoInit with STA on every relevant thread before we get the chance of opening an MTA? This code makes the assumption that won't happen.

Other code should not be creating STAs in threads in general. It's only needed for very specific UI code (e.g. ControllerInterface getting RawInput window messages) that ends up blocking the entire thread (GetMessage/TranslateMessage/DispatchMessage), so it's unlikely that it would be used in one of the main threads. If it is it's probably an error.

This code does make assumptions, but I think they hold up. In general, a thread that gets a cubeb context will exist at least until its finished with the context.

Other options could be making every thread in Dolphin initialize the MTA, or adding a thread to the cubeb context just to keep a reference to the MTA. I think those are overkill.

Reply:
https://docs.microsoft.com/en-us/windows/win32/api/audioclient/nf-audioclient-iaudioclient-initialize
Note  In Windows 8, the first use of IAudioClient to access the audio device should be on the STA thread. Calls from an MTA thread may result in undefined behavior.

Reply:
> Note In Windows 8, the first use of IAudioClient to access the audio device should be on the STA thread. Calls from an MTA thread may result in undefined behavior.

[Firefox noted this a while back](https://bugzilla.mozilla.org/show_bug.cgi?id=1103824) and decided "I don't think we've ever seen evidence that this matters. It's possible it's only relevant in configurations we don't exercise (e.g. exclusive mode)."

Since the market share of Windows 8 is ~5% or less (not sure what it is for Dolphin), and Firefox hasn't seen issues, I don't think it's worth trying to handle that.